### PR TITLE
Feature/immutable

### PR DESCRIPTION
**Pull request by lessthanoptimal**
Thank you for your contribution to the EJML project.

Before submitting this PR, please make sure:
- [x] Your code is covered by tests
- [x] You ran `./gradlew spotlessJavaApply`

### CODE REVIEW

**Comment by lessthanoptimal** [comment]
Previous requests for an ImmutableMatrix https://github.com/lessthanoptimal/ejml/issues/155 . Ok I can find only one, but I know this has come up several times previously.

**Comment by lessthanoptimal** [comment]
@JozsefKutas @ennerf @FlorentinD Does this look good? The PR adds an ImmutableMatrix which is an interface SimpleMatrix implements.

**Comment by lessthanoptimal** [comment]
Might rename it from ImmutableMatrix to ConstMatrix. Const is shorter...

**Comment by JozsefKutas** [comment]
Hi, I've just looked over the changes and I have a few thoughts.

At the moment, the `ImmutableMatrix` isn't actually immutable. A class is immutable if it comes with a gaurantee that instances can't be modified. There's a quick summary of what this implies at https://math.oxford.emory.edu/site/cs171/immutability/ (taken I think from Bloch's Effective Java).

With the current implementation, there's no guarantee that the underlying data of `ImmutableMatrix` will not be modified after the immutable matrix is created:

```java
DMatrixRMaj matrix = CommonOps_DDRM.identity(2);
ImmutableMatrix immutable = SimpleMatrix.wrap(matrix);
matrix.set(1, 1, 0d);
```

It's closer to the unmodifiable collections returned by some methods in `java.util`. Unmodifiability is a weaker condition than immutablility, but is still useful (see https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/util/Collection.html#unmodifiable).

It's not strictly unmodifiable either, because it's possible to mutate elements by downcasting:

```java
ImmutableMatrix immutable = SimpleMatrix.identity(2);
SimpleMatrix simple = (SimpleMatrix) immutable;
simple.set(1, 1, 0d);
```

If you go with this implementation then I'd change the name, just to make it clearer to the user what guarantees the class does and doesn't provide.

Alternative approaches would be:

- Have `ImmutableMatrix` extend `SimpleBase`, and throw an `UnsupportedOperationException` in public methods that mutate the underlying data. Make sure methods like `wrap` make copies. If any public mutator methods in `SimpleBase` are used internally in the implementation of non-mutator methods in `SimpleBase` (to prepare the returned matrix) then this will require workarounds. This is the approach taken by Guava (https://github.com/google/guava/wiki/ImmutableCollectionsExplained).

- Rename the current `ImmutableMatrix` interface, and use it as the parent for mutable matrices (`SimpleMatrix`) and immutable matrices. This is similar to the approach used in scala (https://docs.scala-lang.org/overviews/collections-2.13/overview.html). This approach is more complicated, but it avoids using runtime errors to enforce immutability.

Besides these issues, I don't understand the motiviation behind removing the type parameter from `SimpleBase`. It means errors like this aren't caught until runtime:

```java
ImmutableMatrix matrix = SimpleMatrix.identity(2);
StatisticsMatrix scaled = matrix.scale(2d);
// Exception in thread "main" java.lang.ClassCastException: class org.ejml.simple.SimpleMatrix cannot be cast to class org.ejml.example.StatisticsMatrix ...
```

**Comment by FlorentinD** [comment]
In general I like the idea of having an immutable matrix. One use-case could be for input data which you want to make sure you dont modify by accident.

For the PR itself, I agree with Jozsef and I am also missing some tests which show the Exception is thrown.

Also, starting with the API over the SimpleMatrix makes sense, but I could also see this as useful for more concrete implementations as this is a feature more advanced users would use.

**Comment by lessthanoptimal** [comment]
> Hi, I've just looked over the changes and I have a few thoughts.
> 
> At the moment, the `ImmutableMatrix` isn't actually immutable. A class is immutable if it comes with a gaurantee that instances can't be modified. There's a quick summary of what this implies at https://math.oxford.emory.edu/site/cs171/immutability/ (taken I think from Bloch's Effective Java).

Ok Immutable isn't the best description clearly. It's main design goal is to make it so that someone has to go out of their way to modify it. Kinda of like const in C++, which apparently is considered "a shallow immutable object" as you can strip it of the const property in various ways.

Trying to prevent someone from downcasting and modifying or any other "attack" I think is overkill. This is mainly to keep honest people honest and make it easy to see if someone is deliberately breaking the contract. What I don't like about Unmodifiable is that some of the checks are at runtime, which can result in a nasty surprise.

Maybe call it ConstMatrix?

> * Rename the current `ImmutableMatrix` interface, and use it as the parent for mutable matrices (`SimpleMatrix`) and immutable matrices. This is similar to the approach used in scala (https://docs.scala-lang.org/overviews/collections-2.13/overview.html). This approach is more complicated, but it avoids using runtime errors to enforce immutability.

That's what currently done I believe.

> Besides these issues, I don't understand the motiviation behind removing the type parameter from `SimpleBase`. It means errors like this aren't caught until runtime:

Just reverted that change. It was result of an earlier experiment that didn't go anywhere and I forgot to undo it.

Thanks for all the thoughtful comments!

@JozsefKutas

**Comment by lessthanoptimal** [comment]
See [KalmanFilterSimple](https://github.com/lessthanoptimal/ejml/blob/feature/immutable/examples/src/org/ejml/example/KalmanFilterSimple.java) for an example of how this could be put into practice.

**Comment by lessthanoptimal** [comment]
Unless there's some additional feedback or suggestion for a better name, I'm going to merge this in soon. Planning on doing another release soon too.

File: main/ejml-simple/src/org/ejml/simple/SimpleBase.java [hunk 1]
```diff
@@ -241,20 +210,9 @@ public T kron( T B ) {
...
-     * where c is the returned matrix, a is this matrix, and b is the passed in matrix.
-     * </p>
-     *
-     * @param B m by n matrix. Not modified.
-     * @return The results of this operation.
-     * @see CommonOps_DDRM#mult(DMatrix1Row, DMatrix1Row, DMatrix1Row)
-     */
-    public T plus( T B ) {
+    /** {@inheritDoc} */
+    @Override public T plus( ConstMatrix _B ) {
```
Comment (JozsefKutas) [inline main/ejml-simple/src/org/ejml/simple/SimpleBase.java]:
```suggestion
    @Override public T plus( ConstMatrix<?> _B ) {
```

File: main/ejml-simple/src/org/ejml/simple/SimpleBase.java [hunk 2]
```diff
@@ -373,22 +278,9 @@ public T plusComplex( double real, double imag ) {
...
-     *
-     * <p>NOTE: If the matrix is complex then 'b' will be treated like a complex number with imaginary = 0.</p>
-     *
-     * @param B m by n matrix. Not modified.
-     * @return A matrix that contains the results.
-     * @see CommonOps_DDRM#add(DMatrixD1, double, DMatrixD1, DMatrixD1)
-     */
-    public T plus( double beta, T B ) {
+    /** {@inheritDoc} */
+    @Override public T plus( double beta, ConstMatrix _B ) {
```
Comment (JozsefKutas) [inline main/ejml-simple/src/org/ejml/simple/SimpleBase.java]:
```suggestion
    @Override public T plus( double beta, ConstMatrix<?> _B ) {
```

File: main/ejml-simple/src/org/ejml/simple/SimpleBase.java [hunk 3]
```diff
@@ -398,13 +290,9 @@ public T plus( double beta, T B ) {
...
 
-    /**
-     * Computes the dot product (a.k.a. inner product) between this vector and vector 'v'.
-     *
-     * @param v The second vector in the dot product. Not modified.
-     * @return dot product
-     */
-    public double dot( T v ) {
+    /** {@inheritDoc} */
+    @Override public double dot( ConstMatrix _v ) {
```
Comment (JozsefKutas) [inline main/ejml-simple/src/org/ejml/simple/SimpleBase.java]:
```suggestion
    @Override public double dot( ConstMatrix<?> _v ) {
```

File: main/ejml-simple/src/org/ejml/simple/ConstMatrix.java [hunk 4]
```diff
@@ -0,0 +1,798 @@
...
+    /**
+     * <p>
+     * Returns a matrix which is the result of matrix multiplication:<br>
+     * <br>
+     * c = a * b <br>
+     * <br>
+     * where c is the returned matrix, a is this matrix, and b is the passed in matrix.
+     * </p>
+     *
+     * @param B A matrix that is n by bn. Not modified.
```
Comment (JozsefKutas) [inline main/ejml-simple/src/org/ejml/simple/ConstMatrix.java]:
```suggestion
     * @param B A matrix that is n by p. Not modified.
```

**Review by JozsefKutas — APPROVED** [review]
I'm fine with the name `ConstMatrix`.

A few thoughts:

- The methods `concatRows`, `concatColumns`, `setRow` and `setColumn` should probably take `ConstMatrix` objects as arguments.

- This PR makes it possible to perform operations with different subclasses of `SimpleBase`. This could be useful in some situations, and I don't see any disadvantages. For example, the following code previously wouldn't compile, but now compiles and runs without problem:

```java
DMatrixRMaj matrix = CommonOps_DDRM.identity(2);
SimpleMatrix a = SimpleMatrix.wrap(matrix);
StatisticsMatrix b = StatisticsMatrix.wrap(matrix);
System.out.println(a.plus(b));
```

- Most methods of `SimpleBase` that accept `ConstMatrix` won't work if the user passes an object implementing `ConstMatrix` that doesn't extend `SimpleBase`. You could prevent this by making `ConstMatrix` a sealed interface and making `SimpleBase` a non-sealed class.

I don't think `ConstMatrix` works for my own use case and I'm not completely sure it's worth the additional API complexity, but I don't see any other issues.

**Comment by lessthanoptimal** [comment]
@JozsefKutas Do you know if sealed types are syntax sugar and generate bytecode compatible with Java 11? EJML compiles Java 17 but creates bytecode for Java 11. EJML will be stuck with Java 11 bytecode for the foreseeable future. Took forever to give up on Java 8.

**Comment by lessthanoptimal** [comment]
As a side note, I'm not really the target audience for ConstMatrix either. I mostly work with the low level API. I think it applies to people who attempt to restrict access as much as possible and feel the need to use final ever place possible. I also liked the Equations API but I think I'm the only user of that API after all these years. That is probably a good candidate to be removed in the future, at least if it wouldn't break a bunch of my own code.

**Comment by JozsefKutas** [comment]
> Do you know if sealed types are syntax sugar and generate bytecode compatible with Java 11?

They're not just syntax sugar, so I'd guess they're not compatible.

**Pull request merged**